[JSC] Carefully avoid speculation when it is not necessary in CompareStrictEq

Constellation opened the pull request:
#### 62afbec76ca8be22b3d56e66c61b7e4cc9e590e2
<pre>
[JSC] Carefully avoid speculation when it is not necessary in CompareStrictEq
<a href="https://bugs.webkit.org/show_bug.cgi?id=272937">https://bugs.webkit.org/show_bug.cgi?id=272937</a>
<a href="https://rdar.apple.com/126718615">rdar://126718615</a>

Reviewed by Keith Miller.

We found that sometimes we do speculation too strongly. For example, the code may have an object or undefined,

    if (object.getter !== true)
        ...

In the above case, right now, we put OtherUse on object.getter result when the prediction says it is Other.
But if rhs or lhs is a constant and it is boolean / undefined etc., CompareStrictEq can be performed as a bitwise comparison,
so no speculation is required. If object.getter&apos;s result is used in the other place later, then it will have speculation at
that place anyway. So given that CompareStrictEq does not produce a new prediction from the input (it always returns Boolean),
it can be nice if we can avoid speculation when one of input is a specific constant.

In this patch, we add a code in DFGFixupPhase, which generates (Untyped,Boolean) and (Untyped,Other) pair speculations when
one of the input is constant.

* JSTests/stress/check-true.js: Added.
(test):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupCompareStrictEqAndSameValue):
* Source/JavaScriptCore/dfg/DFGMayExit.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):

Canonical link: <a href="https://commits.webkit.org/277771@main">https://commits.webkit.org/277771@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cfd51e1101d3df7e08f4d8f95d81016355a00ca5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48516 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27727 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51475 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51203 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44581 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33663 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25250 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/39663 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49098 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25403 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41863 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20777 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22881 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/43034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6572 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/41812 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/44820 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43520 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53109 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/48006 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23562 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19877 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46972 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24827 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42067 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45891 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25632 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/55501 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6917 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24550 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11414 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -16492,45 +16489,24 @@ void SpeculativeJIT::compileExtractFromTuple(Node* node)
-    Jump notEqual = branch32(NotEqual, op1.tagGPR(), op2.tagGPR());
-    compare32(Equal, op1.payloadGPR(), op2.payloadGPR(), result.gpr());
+    Jump notEqual = branch32(NotEqual, op1Regs.tagGPR(), op2Regs.tagGPR());
+    compare32(Equal, op1Regs.payloadGPR(), op2Regs.payloadGPR(), result.gpr());
```
OK, here both `JSValueRegs::payloadGPR()` and `JSValueOperand::gpr()` refer the same the gpr.

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
```diff
@@ -2253,6 +2253,8 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
             // If not I might slightly simplify that check.
             if (node->isBinaryUseKind(BooleanUse)
+                || node->isBinaryUseKind(BooleanUse, UntypedUse)
+                || node->isBinaryUseKind(UntypedUse, BooleanUse)
```
Hmm.. how is that possible the node has two same children but with different use kinds?

kmiller68 replied:
This is saying the node is `@A (@B: BooleanUse, @C: UntypedUse)` or `@A (@B: UntypedUse, @C: BooleanUse)`

Constellation replied:
See FixupPhase. We have a case, which only do speculation on the one part of edge.

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -8169,21 +8169,18 @@ bool SpeculativeJIT::compileStrictEq(Node* node)
         || node->isBinaryUseKind(MiscUse, UntypedUse)
         || node->isBinaryUseKind(UntypedUse, MiscUse)) {
-        compileMiscStrictEq(node);
+        compileBitwiseStrictEq(node);
```
So, the rational here is that
```
MiscUse = bool | null | undefined
UntypedUse = object | null | undefined
OtherUse = null | undefined
```
since JSTrue, JSFalse, JSNull and JSUndefined are unique in the implementation, we can just compare the JSValues.

kmiller68 replied:
`MiscUse = Boolean | null | undefined` `UntypedUse` can be anything. But since `true` `false` `null` and `undefined` have exact representations we can do a bitwise comparison.

kmiller68 commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -8169,21 +8169,18 @@ bool SpeculativeJIT::compileStrictEq(Node* node)
 
-    if (node->isBinaryUseKind(OtherUse)
+    if (node->isBinaryUseKind(BooleanUse, UntypedUse)
+        || node->isBinaryUseKind(UntypedUse, BooleanUse)
```
Nit: Maybe we should add a `node->isReflexiveBinaryUseKind(X, Y)` for this?

Constellation replied:
Sounds good. Done.

hyjorc1 replied:
Actually the `reflexive binary relation` on the set `{x, y}` means, each element is related to itself. Given that the minimal reflexive relation on `{x, y}` is `R = {(x, x), (y, y)}`. So, I think the best reasonable name for this case is `isSymmetricBinaryUseKind(x, y)` which is based on the symmetric relation.

kmiller68 replied:
Yeah you're right, symmetric is what I meant

kmiller68 reviewed (approved):
r=me.

Constellation committed:
[JSC] Carefully avoid speculation when it is not necessary in CompareStrictEq
https://bugs.webkit.org/show_bug.cgi?id=272937
rdar://126718615

Reviewed by Keith Miller.

We found that sometimes we do speculation too strongly. For example, the code may have an object or undefined,

    if (object.getter !== true)
        ...

In the above case, right now, we put OtherUse on object.getter result when the prediction says it is Other.
But if rhs or lhs is a constant and it is boolean / undefined etc., CompareStrictEq can be performed as a bitwise comparison,
so no speculation is required. If object.getter's result is used in the other place later, then it will have speculation at
that place anyway. So given that CompareStrictEq does not produce a new prediction from the input (it always returns Boolean),
it can be nice if we can avoid speculation when one of input is a specific constant.

In this patch, we add a code in DFGFixupPhase, which generates (Untyped,Boolean) and (Untyped,Other) pair speculations when
one of the input is constant.

* JSTests/stress/check-true.js: Added.
(test):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupCompareStrictEqAndSameValue):
* Source/JavaScriptCore/dfg/DFGMayExit.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):

Canonical link: https://commits.webkit.org/277771@main
